### PR TITLE
#4939: Add socks proxy tor support to lncli

### PR DESCRIPTION
**Pull request by hieblmi**
This PR adds a `--socksproxy` option to lncli to allow for RPC calls via Tor.

Example: `lncli --rpcserver=remoterpchost.onion:10009 --socksproxy=localhost:9050 --tlscertpath=/cert/path/tls.cert --macaroonpath=/macaroon/path/admin.macaroon fwdinghistory --start_time "-1h"`

I have tested the above example on two physically remote systems and was able to retrieve the forwarding history.

I am grateful for any questions or requests around this code change.

### CODE REVIEW

File: cmd/lncli/main.go [hunk 1]
```diff
@@ -276,6 +293,11 @@ func main() {
 			Usage:     "The path to lnd's base directory.",
 			TakesFile: true,
 		},
+		cli.StringFlag{
+			Name:  "socksproxy",
+			Value: defaultLndDir,
+			Usage: "The host:port of a SOCKS proxy through which all connections to the LN daemon will be established over.",
```
Comment (guggero) [inline cmd/lncli/main.go:299]: nit: break line after 80 characters (counting the tab character as 8 spaces).

**Review by guggero — COMMENTED** [review]
Thanks for the feature!

utACK, only style nits. Please take a look at [our code contribution guideline](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) about how to prefix the commit message.

File: cmd/lncli/main.go (hunk 1 above)
Comment (hieblmi) [reply cmd/lncli/main.go]: @guggero I've updated it according to the other `cli.StringFlag` options. Please let me know if that looks good.

**Comment by 3nprob** [comment]
Related: https://github.com/lightningnetwork/lnd/pull/5651

**Comment by guggero** [comment]
Can you please rebase and squash the commits?

**Comment by hieblmi** [comment]
Hi Oliver, I rebased but missed to squash the commits.

**Comment by guggero** [comment]
You didn't rebase, you merged the master branch into your branch, which created an extra commit. Are you using the GitHub UI only? I think you can't rebase using that.
If you like we can squash the commits when merging, but we only do that for very small PRs that make sense as a single commit.

**Comment by hieblmi** [comment]
Yes I was using the GitHub UI. So should I still rebase and squash this MR in it's current form? I will do that from command line then.

**Comment by guggero** [comment]
> Yes I was using the GitHub UI. So should I still rebase and squash this MR in it's current form? I will do that from command line then.

Yes, that would be great, thanks!

File: cmd/lncli/main.go [hunk 2]
```diff
@@ -276,6 +293,13 @@ func main() {
 			Usage:     "The path to lnd's base directory.",
 			TakesFile: true,
 		},
+		cli.StringFlag{
+			Name:  "socksproxy",
+			Value: defaultLndDir,
```
Comment (guggero) [inline cmd/lncli/main.go:298]: The value is incorrect here. Should be empty!

File: cmd/lncli/main.go [hunk 3]
```diff
@@ -173,10 +176,24 @@ func getClientConn(ctx *cli.Context, skipMacaroons bool) *grpc.ClientConn {
...
+		socksProxy := ctx.GlobalString("socksproxy")
+		torDialer := func(_ context.Context, addr string) (net.Conn, error) {
+			return tor.Dial(
+				addr, socksProxy, false, false,
+				tor.DefaultConnTimeout,
+			)
+		}
+		opts = append(opts, grpc.WithContextDialer(torDialer))
+	} else {
+		// We need to use a custom dialer so we can also connect to unix sockets
```
Comment (guggero) [inline cmd/lncli/main.go:191]: nit: line too long with tabulator width set to 8 spaces.

**Review by guggero — COMMENTED** [review]
Needs a rebase and an entry in the 0.15 release notes.

**Comment by hieblmi** [comment]
Closing this PR and create a new one with rebases instead of extra merge commits.

**Comment by 3nprob** [comment]
@hieblmi For next time, no need to close and reopen, and it's probably preferable to keep the same, to retain review history and comments etc. In case you're not used to this workflow you can do it interactively, for example, when on your branch:

```
$ git remote
hieblmi
origin

$ git rebase -i origin/master
# starts interactive rebase - here you should `drop` the merge commits, and any other commits that are not part of your changes for this PR
# resolve any conflicts, then if necessary:

$ git rebase --continue

# test and inspect

$ git rebase -i origin/master

# now you can `squash` or `fixup` your commits into one - optionally you can just do this all in the first step for smaller PRs where the commits aren't independent

$ git push --force-with-lease hieblmi master
```

**Comment by hieblmi** [comment]
Thank you @3nprob for providing a starter - I am indeed unfamiliar with this flow. I will open a new PR and follow your instructions going forward.

**Comment by 3nprob** [comment]
If you are, I would also suggest to take the opportunity and submit a different branch (e.g. `git checkout -b lncli-socks-proxy` / `git push -u hieblmi lncli-socks-proxy`) rather than submitting your `master`

**Pull request closed without merging**